### ISSUE TITLE
java.lang.InterruptedException when Docker host has high load

### ISSUE THREAD

**Issue by saxonww** [jenkinsci/docker-plugin#640]
Please don't use the issue tracker to ask questions, join jenkins-users mailing list.

If you get some troubles with docker-plugin, please report
 - [x] docker-plugin version you use

1.1.3

 - [x] jenkins version you use

2.89.4

 - [x] docker engine version you use

17.12.0-ce and 18.03.0-ce

 - [x] stack trace / logs / any technical details that could help diagnose this issue

We are seeing odd behavior with some of our builds where they will occasionally loop indefinitely. In all cases these are Angular builds using the Angular CLI v1.7.1.

Whenever we see a looping build, the console output will show that Jenkins cannot contact the Docker container:

```
[Test] Cannot contact docker-73d3b2b2679a2: java.lang.InterruptedException
```

After this point, the output of the currently-running command appears to be rewritten to the console, over and over again, until the job times out. The output looks identical - timestamps and durations printed by the tools are identical with each copy of the output. Once the build times out, we finally get some different output:

```
[Test] Cancelling nested steps due to timeout
[Test] Could not connect to docker-73d3b2b2679a2 to send interrupt signal to process
[Pipeline] [Test] }
[Pipeline] [Test] // timeout
[Pipeline] [Test] }
[Pipeline] [Test] // withEnv
Post stage
[Pipeline] [Test] junit
[Test] ERROR: Issue with creating launcher for agent docker-73d3b2b2679a2. The agent has not been fully initialized yet
[Pipeline] [Test] }
[Pipeline] [Test] // node
[Pipeline] [Test] }
[Pipeline] [Test] // stage
[Pipeline] [Test] }
[Test] Failed in branch Test
[Pipeline] // parallel
[Pipeline] }
[Pipeline] // stage
```

This ends up being a snowball problem for us, because developers will go and retrigger their build, and if this is load-related it just ends up exacerbating the issue. We have dialed down the number of containers allowed to run on each host (16 CPU/256GB RAM) to 4, and at most we're kicking off 2 processes per build (node + chrome), so we are at this point wondering if this is really load related or something else.

We are using the 'Attach Docker Container' connect method. We're not sure if SSH-based or JNLP-based would show different/better behavior. We have not set any memory or CPU limits on these containers so far.

**Comment by saxonww** [comment]
Update on this: we've made changes to our build which reduce the amount of generated console output; we were running angular builds with karma coverage reports dumping to the console. It's hard to draw good conclusions from this, but every change submitted *after* reducing the console output has built fine. Older changes which had not been rebased yet (i.e. still generating extra console output) continue to exhibit the timeout and looping behavior, even when no other builds are running (i.e. no load). Does this make any sense? Is it possible that the Jenkins -> Docker container connection could be affected by console logging behavior?

**Comment by pjdarton** [comment]
The direct docker connection method is still experimental (it's fairly new; SSH and JNLP are old) so it is possible.

**Comment by pjdarton** [comment]
There's been a new version of the plugin released (1.1.5).  If you could update to that and see if it's any better for this issue, that'd be appreciated.
FYI the new plugin has improved JNLP support so, if the direct-connection method is under suspicion, you should be able to reconfigure things to use a (custom) JNLP connection instead without having to do anything to your docker image.
